### ISSUE TITLE
What is an Inbox for?

### ISSUE THREAD

**Issue by DavidStaab** [featherweight/featherweight#44]
These are open-ended questions that might shed a lot of light on one of the fundamental assumptions FTW appears to make about Actors: Every actor comes with a predefined request-type message interface when it gets launched.

1. I understand that the Inbox is a parallelized REP socket with some extra features and logic bundled in. Can you enumerate those features and their utility in FTW?
2. Why not just create an Actor template that has a SyncServer loop for inbound messages and a separate job-worker loop? Why abstract the message processing loop into a dynamically launched VI that the framework maintains on its own?
3. Are there any known situations where a developer wants to customize an actor's Inbox but can't because it's baked into the framework?
4. Is it reasonably straightforward to create an Actor and attach a SyncServer or Subscriber to it alongside the Inbox? Is it reasonable to destroy the Inbox event registration when the actor launches and use a different messaging paradigm like Subscriber, instead?

**Comment by wirebirdlabs** [comment]
Great questions ... I've edited the comment in order to number the questions to answer them individually. It's after hours here currently, but stay tuned for answers to roll in over the next day ...

**Comment by wirebirdlabs** [comment]
2. **Short answer:** a blocking `Call Library Function Node` would be required. **Longer answer:** A blocking CLFN will block the thread it's running in. A LabVIEW execution system has several thread pools based on a combination of both an arbitrary affinity group and a priority level. Nominally, each thread pool is initialized with 20 threads.

A while back, while architecting a larger system, we started noticing random system behavior as more and more actors came online. It was incredibly difficult to pinpoint what would fail and when. Fast forward a few weeks of troubleshooting and red herrings -- turns out we were running into thread pool exhaustion. It appears that the scheduler for execution systems will just silently, patiently wait to retry task execution when all threads are currently blocked -- anecdotally, it may have even been throwing away some clumps of code that couldn't be scheduled immediately.

The lesson learned was this -- blocking CLFN's can pile up in a system much sooner than you might expect -- around the 12-14 currently-running actors mark -- and the failure mode from this happening is baffling and incredibly hard to pinpoint.

After that point, the Inbox was redesigned to use User Events posted by `libftw.*`. This allowed us on the LabVIEW-side to use its own mechanism for multiplexing incoming signals -- the Event Callback Registration. Here, some type of fan-in enables far more blocking sockets than the available thread pool.

Fast forward to early 2016 -- a bug was found with Event Callbacks (CAR 567875), which should be fixed in some forthcoming LabVIEW release. In the meantime, FTW has been re-architected to use the Event Handler Structure rather than Event Callbacks, which incidentally flushed another bug in LabVIEW with User Events (CAR 567877).

So ... the Inbox has an illustrious history of pushing the boundaries of asynchronous communication from a shared library into LabVIEW, and for better or worse most of the development time has gone toward debugging some pretty cryptic failure modes. I look forward to spending more time on better documentation between a SyncServer and an Inbox, but hopefully this helps delineate the difference between the two. (spoiler: the SyncServer *might* be going away altogether, and if it sticks around, will at least change names in the future)

**Comment by wirebirdlabs** [comment]
3) No known problems so far using Inbox with another reference design template for actors. As one point of reference, check out `ftw-master\tests\core\FTW-Test-Socket-Stress-InboxConnector.vi` which demonstrates "naked" calls to Inbox.

**Comment by DavidStaab** [comment]
`5. Is a Connector (FTW-Connector-Construct.vi) the REQ counterpart to an Inbox's REP?

**Comment by wirebirdlabs** [comment]
5) Yes.

Expanding, a `FTW Actor Connector` is named a Connector because it connects to an Actor Instance. Analogous, like a Database Connector is the client side code that connects to a Database Server Instance.

There was a long history of what this feature should be called -- Actor API (since it's the API of the actor), Actor Reference (since, it's oftentimes an appropriate mental model to think of an Actor Instance as a by-ref object, on which remote procedure calls are invoked by a reference to the object), and more such as Actor Client, Actor Binding, Actor Alias ...

Essentially, the FTW Actor Connector is just a language binding that wraps the implementation of interfacing with the Actor Instance. Incidentally, it may be in the same language as the instance; incidentally, they may even reside on the same machine and even in the same process.

There exist zero static linkages between the Connector and the Instance, since a remote connection to the actor should not be linked to the business logic of the actor. Just as a client-side webpage may be served via a backend database, the client does not need to have that database installed (since the output of that data source was serialized to a TCP/IP stream). The "linkage" between the FTW Actor Connector and the FTW Actor Instance is merely the transport layer itself -- typically, the network stack, pipes in the kernel, or even in-process memory exchange.

Oh, and the only other linkage is the format of the messages themselves. Effectively, this is just a written contract between the two, yet importantly, **not enforced by the type system**. For convenience, messages -- Requests, more specifically -- may be constructed as type definitions within the language -- to provide a level of type safety between the Instance and the Connector.

Yet, a Request structure is orthogonal to even the Instance and the Connector. `Actor X` could implement `Messages A and B and C`, and `Actor Y` could implement `Messages B and C and D`. `Connector C` could wrap `Messages B and C` and effectively be a Connector into both `Actor X` and `Actor Y`. `Connector D` could then wrap `Messages A and B` and only be intended for `Actor X` but not `Actor Y`.

It's typical to have a 1:1 relationship between `FTW Actor Instance` and `FTW Actor Connector`, but not required.

Messages themselves, if implemented as a type definition, should probably never be statically linked to or "owned by" the Instance, because a Connector linking to that type would undesirably link to the Instance.

**Comment by wirebirdlabs** [comment]
4a) Sure; again, check out: `ftw-master\tests\core\FTW-Test-Socket-Stress-InboxConnector.vi`
4b) An Actor Instance is not an actor without an Inbox, in FTW anyway. The only other atom of concurrency is a `Worker` in FTW. Currently, this does not have a real definition in FTW, because its structure and purpose can be so varied, but it can be loosely defined as "anything that runs concurrently but is not an Actor". As an example, `ftw-master\tests\core\FTW-Test-Socket-Stress-PubSub.vi` demonstrates what we could call two Workers within the same block diagram.

**Comment by wirebirdlabs** [comment]
2a) Blocking CLFN -- see above.
2b) See answer to 1) below

**Comment by wirebirdlabs** [comment]
1) The `FTW Actor Inbox` is the front line of defense between "the outside world" and the `FTW Actor Instance`.

Recall, a mental model for the underlying interface into an Actor is a byte stream on a socket -- as compared to a type-safe ABI that has been verified by a compiler.

Further, one tenet of actor-oriented programming is an actor's responsibility to maintain its own health. Keep doing the core job it's intended to do.

For this reason, the run-time attack vector of "any arbitrary bytes coming across a socket into the Actor Instance" is something we'd like to protect against at the framework layer -- rather than having contingencies within every since Actor Instance as to "how to deal with this sort of message".

When we're talking about this attack vector, it's not typically in the context of a malicious attack -- most of the times, an unintentional programming error. Sending the wrong message to an Actor; sending malformed JSON; sending too many messages too quickly.

So, **the first thing an Inbox does is provide baseline runtime error checking and prevention for the Actor Instance**. The Inbox is able to auto-respond back to the source of the message with a "malformed message" error, without having brokered that malformed message to the Instance. Since again, the Instance's job is to stay healthy, and if it's interrupted just to return an error to the remote client, it's not doing its job.

The `FTW Actor Inbox` has not always been called this. At one point, it was the Message Broker or the Message Relay (since it brokers/relays messages from "the outside world" into the Actor Instance), the Message Scheduler (since it schedules delivery or non-delivery of the message into the Actor Instance), and my favorite is the Bouncer (like the big guy who sits outside the bar).

As such, **the second thing the Actor Inbox does is defines the run-time "API" of the Actor**, and the delivery semantics behind each message type. You can specify which messages are delivered to the actor instance, and which ones bounce back with an error message without bothering the actor.

The **third thing an Inbox provides is the ability to have separate endpoints to which Connectors can connect**. This one is very cool, since it also allows you to prioritize different endpoints. For example, you may have one inbox for "low priority" WebSocket clients who only have access to X subset of "harmless" messages, and then a "high priority" in-process endpoint for critical business logic between the instance and other instances within the application with a larger subset of "not harmless" messages. Importantly, this removes head-of-line blocking for messages traveling on the higher-priority channels compared to the incoming messages on the lower-priority channels.

So, an Inbox may have N endpoints, each with a separate priority level, and responding to subsets of messages.

**The fourth thing an Actor Inbox provides is the ability for the Instance to define timeouts on a per-message definition**. We already know Connectors can define how long they are willing to wait in order to receive a response -- but the Inbox provides the Actor Instance the ability to define how long the Inbox will wait to deliver each message type to the Instance.

This enables you to tune "less important" messages to time out faster unless they're able to be serviced immediately. A good example is querying the status of an Actor to display on a user interface. You may set the delivery timeout even down to 0msec, meaning unless the inbox is able to **immediately** broker that message to the instance since the instance has zero outstanding requests and is ready right now to handle the next request, then the Actor Inbox would reply back to the requestor with the "Actor is Busy" error. Since even this error message represents information, which is better than "the lack of information" during the indeterminate amount of time leading up to a client-side timeout.

And again, being able to set that timeout is all about system health. If there exists a message type that you, as a developer, assert should *always* be able to be serviced immediately (e.g., based on design of the entire system), setting a very low or 0msec timeout provides such a run-time assertion that would be evidenced client-side, and the health of the system would be evidenced in a way that could be observed.

Oh, finally -- this term "pushback" -- that's when a connection becomes congested for whatever reason -- normally due to the Actor Instance not being able to process and respond to queued messages quickly enough -- "pushback" is able to synchronously throw an error on the client-side from a very low layer in the transport layer down in kernel space. When the `recv` buffer of a socket is full on the Actor Inbox bound endpoint, TCP/IP flow control takes over and responds with `EAGAIN` or similar on the Actor Connector connected endpoint. Just as an Inbox protects an Instance from getting bogged down, the TCP/IP stack can protect the Inbox from getting bogged down. So **the fifth thing an Inbox does is defines how many outstanding incoming messages it's willing to buffer; said another way, the amount of physical memory it's willing to set aside on its own computing target to store these until the Actor Instance is able to process them."

There are deep technical reasons why the Inbox must be a separate loop from the Actor Instance in LabVIEW. The current design exists as the only sweet spot between threading and I/O multiplexing models for the scheduler in the NI LabVIEW run-time environment, and how that interacts with the foreign-function interface and threading model in `ftw-kernel`.

And **a sixth thing is defining a per-message length limit that will automatically be rejected and thrown back to the Connector as a protocol violation error**. If your Actor Instance deals with very small messages -- 10's or 100's of bytes worth of JSON on the wire -- then set the message size limit low as well.

If for no other reason, the Inbox is designed to evidence failure as early as possible, to make debugging more sane.

**Comment by wirebirdlabs** [comment]
The final thing to note -- Inbox Configuration is probably the most advanced topic available in Featherweight in order to make the most robust systems imaginable. Fortunately, it ships with sane defaults, where the vast majority of applications should find "everything just works" without having to know any of these things.

Additionally, it's one of the newest concepts, and is therefore particular open to suggestions and able to change.

**Comment by coffeegoat** [comment]
Just to chime in quickly - Hardware system developer here, and a bunch of
these items are above my head *but* the idea of the Inbox bouncing toxic
message is vitally important to me as a hardware developer.  Depending on
how reliable/robust the end node hardware is, I've run into situations
where it is just too stupid to take care of itself.  One such example
resulted in the failure of a ~$1M dollar hardware component.  While
screening this toxic message (it was actually a null character) could have
been done at the actor level, this sort of message should have never
actually made it through the the actor as it was an invalid/malformed
message.

**Comment by omarmussa** [comment]
I'd also like to chime in that one of the beautiful things about the
ActorConnector separation from the Actor itself is that it means that you
can seamlessly/transparently move the Actor via changing the connection
string.  This means an actor can be designed to run on a desktop machine
and then moved to an RT target (with refactoring if needed to handle
inherent RT limitations) and the code that calls the actor connector
doesn't have to change at all (because the only thing that changes is the
actor's address string).  Your application won't have dependencies on the
actor itself so your application code remains thin.

Thank you,

*Omar Mussa*
Director of Software Engineering, EUV Tech       mobile 925 270 8582   |
office 925 229 4388   |   *euvtech.com <http://euvtech.com>*

**Comment by wirebirdlabs** [comment]
> seamlessly/transparently move the Actor via changing the connection string

Exactly -- "location transparency" -- another tenet of Actor-Oriented Programming. This is important to maintain, because sometimes for an Actor to best do its job of maintaining health in a critical system, the Actor itself needs to move to a new compute target. Or new language (e.g., embedded Python).

Or just as plain-ol' run-time dependency injection. Usually, when we say "Dependency Injection", we're talking about the type-safe OO concept -- which has its uses, usually bounded within the context of a single language and run-time environment and type system. But run-time configuration of Actors in this manner enables distributed system composition in a way that object-oriented dependency injection is not quite as suited for handling. _(Both still have their merits and better-suited use cases; I'm only making this comparison because it's a worthwhile consideration seeing run-time Inbox Configuration as what could be a good alternative to one existing method we're familiar with for composing systems at run-time)._

**Comment by wirebirdlabs** [comment]
> the idea of the Inbox bouncing toxic message is vitally important to me as a hardware developer

That's absolutely the idea here -- filter the failure mode as low in the stack as possible. Said another way, as close to the *actual* failure as possible.

When a Connector sends a message to an Instance, it goes through many many layers -- software and phyiscal. If that Connector is sending a bad message, or some other sort of message that just won't succeed for some reason -- we want the failure to be reported and bounced back as early as possible.

This means 1) fewer layers/links to debug in order to trace the original source of the failure mode, and 2) less opportunity for cascaded failure, as the bogus message is transported wholesale all the way to the place where it could (probably) cause the most damage -- the running process of the Actor Instance itself.

**Comment by DavidStaab** [comment]
Here's an attempt to extract the "what" from the "why" in all Jack's replies and summarize the information succinctly:

**Inbox:**
- Provides error checking for inbound messages
  - Auto-responds to the sender if message is malformed
- Provides predefined comm API to actor
  - Message firewall
  - "Poison pill" to auto-kill actor if another entity tells it to (which entities? where is the access control list for poison pill?)
- Multiplicity allows for priority message queuing
- Provides send timeout
- Enforces input buffer size limit (where?)
- Enforces message size limit

And a diagram of the entities involve in FTW's design (or use case) for an Actor:

![default_ftw_actor](https://cloud.githubusercontent.com/assets/1372998/13993850/10280c52-f0f0-11e5-8bb8-220771c7ea7f.png)

**Comment by wirebirdlabs** [comment]
Got it!

> "Poison pill" to auto-kill actor if another entity tells it to

One important semantic here -- a Poison Pill does not have to be swallowed by the Actor. It can refuse. Said another way -- it's not possible to reliably remotely kill an Actor Instance in Featherweight, unless the Instance provides a reliable way to kill itself remotely.

For this reason, that message is just like any other message -- the Access Control List is set as with any other message.

All said ... now diving into code to verify your model, I've misrepresented the current implementation of `ftw-master\src\socket\inbox\core\FTW-Inbox-Core-Configuration.vi`. Although it could be improved to provide per-endpoint access control of messages, right now, access control is a single configuration that applies to all endpoints.

Jogging my memory, the rationale behind this "less powerful" configuration was a mixture of 1) finite amount of time to develop the feature, 2) wanting to get at least some Inbox features out with field validation prior to going "all the way", and 3) simplifying the problem domain of configuration, since it would be required to normalize the data structure -- i.e., the endpoints would need to be enumerated as an iterable array, then all the messages, then a lookup table that binds each message to each endpoint type. Totally do-able, but just felt like such intricate configuration would be a turn-off or too high an initial barrier. Feedback welcome.

**Comment by DavidStaab** [comment]
The Actor needs context in order to decide whether to swallow the pill. Who sent it? Why? If my doctor gives me a pill, I swallow it. If Morpheus gives me a pill, I think twice.

**Comment by wirebirdlabs** [comment]
I fully agree, but have held off intentionally from implementing this in the framework so far, because the most robust solution (end-to-end encryption, or at least signing) has been out of scope up through this time. That said, verifying the endpoint that the message came across -- that could be a simpler problem, but would still take some work based on the current architecture of libnanomsg.

At least within the application space I've worked most, this highest level of security and robustness has not been necessary. It boils down to this -- as a developer, I don't accidentally sprinkle Poison Pill messages in unintentional places -- in other words, it's incredibly unlikely to create that sort of bug. Other sorts of malformed message bugs -- sure, that happens -- but not this particular one. Even if it did happen, it'd be easy to figure out what happened and fix it. And the amount of effort necessary for a non-developer user of the system to craft such a message is somewhere between absurdly improbably to nigh impossible.

Now, that's for embedded systems behind controlled firewalls with physical access security measures at the facilities.

Anything that exposes something to the internet -- this is always done behind a reverse proxy, and never directly to any portion of the LV Run-Time environment itself. I would strongly recommend against, say, creating a `tcp://` or `ws://` endpoint to one of your actors where that computer and port was open on the firewall and internet accessible.

So, agree with you completely, but the risk profile for this specific type of failure has thus far been far below the effort to protect against it.

**Comment by wirebirdlabs** [comment]
By the way -- I'm 100% inclined to accept Pull Requests for these types of features that don't yet exist!

**Comment by DavidStaab** [comment]
So Inbox's implementation is so hairy (across both LV and C, compared to Sync Server or Pub/Sub) because it has to side-step the CLFN blocking behavior to allow lots and lots of LV actors. It does this by running a worker thread in C that actually receives messages, then passes them up to the LV event handler for processing. Subscriber has the same implementation to this point.
- Why not create the same workaround for Sync Server?

Replies from the actor are generated through a CLFN call, though...so that would seem to create the threadlock issue again.
- Does the CLFN call to `ftw_socket_inbox_reply()` indeed threaten to cause threadlock?
  - And Publisher's call to `ftw_publisher_publish()`?

**Comment by wirebirdlabs** [comment]
> Why not create the same workaround for Sync Server?

Briefly, a SyncServer (synchronous server; again, not crazy about this nomenclature) is not designed for scaling in a FTW application within LV.

The only place it is used in the FTW framework right now is during the synchronous handshake when a new actor instance is spawned. This even is slated for replacement, because when massively launching Actors in parallel, it's indeed possible to exhaust an execution system.

It also exists as a low-barrier convenience for interoperability with other systems, since it's a thin wrapper around a plain ol' [REP socket](http://nanomsg.org/v0.8/nn_reqrep.7.html). Whereas the Inbox bakes in some semantics, such as JSON wire protocol ...

Conversely, the Inbox is meant to scale to larger FTW applications. The complexity you mention in Inbox is incidental complexity -- not inherent complexity -- driven by the bogusness of facilities for asynchronously passing data back into LabVIEW without exhausting the scheduler. This complexity is justified, however, because without it, we could not effectively create applications beyond a dozen or so actor instances, without very carefully segregating them into pools based on execution system (that type of system configuration is tedious, and even dangerous. Dangerous, because it's soooo hard to detect failure modes if you're accidentally exhausting an execution system. It's like troubleshooting a glitch in the matrix, blindfolded.)



> Does the CLFN call to ftw_socket_inbox_reply() indeed threaten to cause threadlock? And Publisher's call to ftw_publisher_publish()?

Threadlock makes sure bolts don't come loose :-) But thread exhaustion (or starvation) of a threadpool can totally happen when too many of those calls are happening at once.

Generalizing: all CLFN's block the one worker thread in which they are scheduled for the entire duration of execution (plus whatever bookending the LV execution system puts on that scheduled piece of work). For that matter, so would any clump of LV code. But the thing we tend to get concerned over is this: how long does any one unit of such scheduled work block other queued units of scheduled work? Additionally, a thing we're concerned about is the nondeterminism of execution time -- especially in situations where unbounded nondeterminism is possible -- such as, when the scheduled clump has a spinlock waiting on a mutex held by another thread. Unbounded locking is deadlock, which means your worker pool is effectively down a player (that's pretty much always a bad thing).

Thread exhaustion is not *bad* all the time -- it's only bad if it leads to something bad happening. Like if there are 20 actors in a 20 thread worker pool -- as long as at any given time, 10 are `recv()`ing and 10 are `send()`ing, presumably it'll all continue to run (albeit at 50% throughput). But if somehow all 20 `send()`s are scheduled concurrently, and all 20 `recv()`s are waiting in the queue to be scheduled -- oops, your system is deadlocked.

Thread exhaustion can actually be a *good* thing, when you're running a computationally-intensive application where the sweet spot for goodput of calculation is greater than the number of cores on the machine, but less than infinity threads (because there exists overhead with synchronizing I/O of threads; for each additional unit of concurrency you introduce, the CPU is doing less real work and more bookkeeping).

Moving back to `ftw_socket_inbox_reply()` for a moment. Consider an execution system that always has enough actors running in it that at any given time we're 90% likely to have exhausted our execution system default of 20 worker threads (i.e., if we set a breakpoint pausing all threads we would see nominally all 20 threads with `ftw_socket_inbox_reply()` in their call stack).

This is one of the benefits of having the [FTW Actor Inbox own its own dedicated thread](https://github.com/featherweight/ftw-kernel/blob/master/framework/ftw_framework.c#L177-L181) running in the FTW Kernel, rather than squatting on one of the worker threads in a LV execution system. Because, the "perfect" `recv()` here would wait exactly 100% of the time -- the recv process would be instantaneous, and it'd be back to waiting again for the next message.

Consuming a thread for such a task is exorbitantly expensive, and a non-starter for non-trivial applications with more than a few actors. Sparing the details, the design of the Inbox is closer to constructing an interrupt that then schedules a finite piece of work to consume a worker thread's time when a message actually comes in.

This is a lot of rambling, but does it help?

**Comment by DavidStaab** [comment]
> This is a lot of rambling, but does it help?

Honestly, rarely. It's hard work parsing the value (or just the on-topic content) out of the rambles. I'm tempted to just write a lexer+parser for Jack-speak. Nonetheless, let's dig in:

> [Sync Server] exists as a low-barrier convenience for interoperability with other systems, since it's a thin wrapper around a plain ol' REP socket. Whereas the Inbox bakes in some semantics...

Yes, but I'm not interested in those semantics. My question is about the design of _safely_ passing messages between nanomsg and a LV actor. **Ignoring the "framework" parts of an Inbox, why not create the same safety on a LV REP socket as you've created on an Inbox?** Then, if someone needs a nanomsg REP actor in LV, they get one without worries. If they need an FTW (framework) actor, they also get that without worries.

> ...[The Inbox's] complexity is justified...

Yes, I agree. There's no need (between us) to justify it further.

> [Threadlock vs. thread exhaustion]

Sorry, I chose the wrong word. You correctly guessed the concept I meant to reference, though.

I read between the lines that the reason thread exhaustion is a problem for FTW's actors is that you sometimes want to spin a bunch of them up _simultaneously_, and it's exactly (**and only?**) that situation that creates deadlock or other weirdness as a result of thread exhaustion. Is that right?

**Comment by wirebirdlabs** [comment]
> Ignoring the "framework" parts of an Inbox, why not create the same safety on a LV REP socket as you've created on an Inbox?

The Inbox is very closely tied to things expected by existing actor reference design: https://github.com/featherweight/featherweight/blob/expressions/templates/FTW-ActorTemplate.vi

An actor written in another language only needs to abide by the wire-level contract to communicate with an Inbox: a nanomsg REP socket, sending JSON.

The SyncServer is a thin wrapper around REP; the wire format can be anything. The only thing in FTW that uses this is the handshaking when spawning a new LV actor from LV, but it probably won't use this forever (because of the thread exhaustion issues).

**Comment by DavidStaab** [comment]
I think you're still missing my point: I'm not asking for Sync Server to look like Inbox; I'm asking for _one_ subset of Inbox's features, a non-thread-exhausting worker that applies the back-pressure that would be applied by a REP socket written in C, to be bolted onto Sync Server, as well. I see no reason for letting Sync Server fall victim to LV's limitations while Inbox doesn't.

The other features of Inbox that pertain to your framework can stay with Inbox; they distinguish it from nanomsg's REQREP and create the need for the framework, in the first place.

**Comment by wirebirdlabs** [comment]
> I see no reason for letting Sync Server fall victim to LV's limitations while Inbox doesn't

Aha; the key distinction here is simply that Sync Server was not originally intended as an extension point in the manner you're considering. There's no reason it's technically infeasible -- it just has not been on the radar.

That said, it is CERTAINLY acting as a potential bottleneck in its only current role within FTW, and that is as the temporary socket used to handshake when asynchronously spawning a new Actor Instance.

And speaking of asynchronously spawning new actor instances (see next comment) ...

**Comment by wirebirdlabs** [comment]
... not all Actor Instances must be spawned in FTW. They don't even have to be reentrant. It can sit as a static call on a top-level VI. This of course doesn't scale if your system requires scaling out to N run-time instances, but it does explicitly allow a much better debugging experience with a simpler system if your application domain is inherently static anyway.

The reason I bring this up is 1) just because it's useful, but also 2) it's relevant to this SyncServer handshaking -- since the handshaking does not happen. The handshake only happens with spawned instances, not statically-called instances.

How is this done? When spawned, the spawner writes the JSON element `FTW-LauncherAddress` into the Actor Instance configuration, alerting the new instance that it needs to handshake with its spawner. On the other hand, a statically-called instance nominally has nothing writing this element, so it will not attempt to handshake (it's cased-out as a no-op).

**Comment by wirebirdlabs** [comment]
@DavidStaab, that bit of information about statically-calling Actor Instances is especially useful for real-time developers.
